official/utils/testing/golden.py

```diff
@@ -0,0 +1,252 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""TensorFlow testing subclass to automate numerical testing.
+
+  Golden tests determine when behavior deviates from some "gold standard", and
+are useful for determining when layer definitions have changed without
+performing full regression testing, which is generally prohibitive. This class
+handles the symbolic graph comparison as well as loading weights to avoid
+relying on random number generation, which can change.
+
+The tests performed by this class are:
+
+1) Compare a generated graph against a reference graph. Differences are not
+   necessarily fatal.
+2) Attempt to load known weights for the graph. If this step succeeds but
+   changes are present in the graph, a warning is issued but does not raise
+   an exception.
+3) Perform a calculation and compare the result to a reference value.
+
+This class also provides a method to generate reference data.
+
+Note:
+  The test class is responsible for fixing the random seed during graph
+  definition. A convenience method name_to_seed() is provided to make this
+  process easier.
+
+The test class should also define a .regenerate() class method which (usually)
+just calls the op definition function with test=False for all relevant tests.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+
+import argparse
+import hashlib
+import json
+import os
+import shutil
+import warnings
+
+
+import numpy as np
+import tensorflow as tf
+from tensorflow.python import pywrap_tensorflow
+
+
+class BaseTest(tf.test.TestCase):
+  """TestCase subclass for performing golden tests.
+  """
+
+  def regenerate(self):
+    """Subclasses should override this class to generate new reference data.
+    """
+    raise NotImplementedError
+
+  @property
+  def file(self):
+    """Subclasses should override to "return __file__".
+    """
+    raise NotImplementedError
+
+  @property
+  def data_root(self):
+    """Use the subclass directory rather than the parent directory.
+
+    Returns:
+      The path prefix for reference data.
+    """
+    return os.path.join(os.path.split(
+        os.path.abspath(self.file))[0], "test_data")
+
+  @property
+  def ckpt_prefix(self):
+    return "model.ckpt"
+
+  @staticmethod
+  def name_to_seed(name):
+    seed = hashlib.md5(name.encode("utf-8")).hexdigest()
+    return int(seed, 16) % (2**32 - 1)
+
+  @staticmethod
+  def common_tensor_properties(input_array):
+    """Convenience function for matrix testing.
+
+    Args:
+      input_array: Tensor (numpy array), from which key values are extracted.
```

Comment:
It's tricky because it's not a tf.Tensor, it's the result of .eval() being called and is therefore a numpy array. I thought if I called it tensor it would imply it was an instance of tf.Tensor.